R/cv_sgdnet.R

```diff
@@ -0,0 +1,253 @@
+#'   `name`       \tab the type of prediction error used\cr
+#'   `fit`        \tab a fit from [sgdnet()] to the full data set based on the
+#'                     `alpha` with the best cross-validation score\cr
+#'   `alpha_min`  \tab the `alpha` corresponding to the fit with the best
+#'                     cross-validation performance\cr
+#'   `lambda_min` \tab the `lambda` corresponding to the fit with the best
+#'                     cross-validation performance\cr
+#'   `lambda_1se` \tab the largest `lambda` with a cross-validation performance
+#'                     within one standard deviation of the one
+#'                     coresponding to `lambda_min`
+#' }
+#'
+#' @seealso [sgdnet()], [predict.cv_sgdnet()], [plot.cv_sgdnet()]
+#'
+#' @export
+#'
+#' @examples
+#' set.seed(1)
+#' n <- nrow(heart$x)
+#' train_ind <- sample(n, floor(0.8*n))
+#' cv_fit <- cv_sgdnet(heart$x[train_ind, ],
+#'                     heart$y[train_ind],
+#'                     family = "binomial",
+#'                     nfolds = 7,
+#'                     alpha = c(0, 1))
+#' plot(cv_fit)
+#' predict(cv_fit, heart$x[-train_ind, ], s = "lambda_min")
+cv_sgdnet <- function(x,
+                      y,
+                      alpha = 1,
+                      lambda = NULL,
+                      nfolds = 10,
+                      foldid = NULL,
+                      type.measure = c("deviance",
+                                       "mse",
+                                       "mae",
+                                       "class",
+                                       "auc"),
+                      ...) {
+
+  stopifnot(nfolds > 2, is.numeric(alpha), length(alpha) > 0)
+
+  type.measure <- match.arg(type.measure)
+
+  x <- as.matrix(x)
+  y <- as.matrix(y)
+
+  n_samples <- nrow(x)
+  n_features <- ncol(x)
+  n_alpha <- length(alpha)
+
+  if (!is.null(nfolds)) {
+    if (nfolds > n_samples)
+      stop("you cannot have more folds than samples.")
+  }
+
+  if (is.list(lambda)) {
+
+    if (n_alpha != length(lambda))
+      stop("the length of the lambda list needs to match the number of alpha.")
+
+  } else if (is.vector(lambda)) {
+
+    if (n_alpha > 1)
+      stop("you need a list of lambdas (or have it set at NULL) when you have multiple alphas.")
+
+    lambda <- list(lambda)
+
+  } else if (is.null(lambda)) {
+
+    lambda <- replicate(n_alpha, c(NULL))
+
+  }
+
+  fits <- vector("list", n_alpha)
+  for (i in seq_along(alpha))
+    fits[[i]] <- sgdnet(x, y, lambda = lambda[[i]], alpha = alpha[i], ...)
+
+  lambda <- lapply(fits, "[[", "lambda")
+
+  nlambda <- vapply(fits, function(x) length(x$lambda), double(1))
+
+  if (is.null(foldid))
+    foldid <- as.numeric(cut(sample(n_samples), nfolds))
+  else {
+    if (length(foldid) != n_samples)
+      stop("the length of `foldid` must match the number of samples")
+    nfolds <- length(unique(foldid))
+  }
+
+  cv_raw <- vector("list", n_alpha)
+
+  for (i in seq_along(alpha)) {
```

Comment:
Would it be possible to parallelize this using `foreach`? I'm not sure it would be worth it for most problems (especially if we are as fast as we hope to be) but seems like it might be useful.

Reply:
Yeah, that would probably not be any problem. My reason for not including this straight up is that if we parallelize `sgdnet()`, there is no benefit from also having `cv_sgndet()` parallelized.

Reply:
I'll save this for later in any case.